rsdockerspawner/rsdockerspawner.py

```diff
@@ -621,6 +625,17 @@ def __slot_free(self):
         self.__slot = None
         self.__pools_dump()
 
+    @classmethod
+    def __slots_from_dump_file(cls):
+        p = pkio.py_path(_POOLS_DUMP_FILE)
+        if not p.exists():
+            return PKDict()
+        slots = PKDict()
+        for v in pkjson.load_any(p).values():
+            if isinstance(v, dict) and 'slots' in v:
+                for s in v.slots:
```

Comment:
I think you need to be careful here. The previous slots are bound to a pool, not a cname. A container may be named, e.g. jupyter-robnagler, but it could be in the wrong pool due to a config change between starts. This is a not unusual occurrence, e.g. moving students to a class pool and then returning them back to the community pool. 

Therefore, previous_slots should be previous_pools, which has slots.

Reply:
> The previous slots are bound to a pool, not a cname.

But a cname is bound to a slot and I care about the details of a slot (`activity_secs` and `start_time`) for a specific cname. I'm having a hard time understanding how pools are relevant?

> A container may be named, e.g. jupyter-robnagler, but it could be in the wrong pool due to a config change between starts.

True but I don't believe that has an impact on the times which is the only pieces of data I'm after. Regardless of whether or not a container is in the wrong pool after a restart it's `activity_secs` and `start_time` remain the same. 

Maybe I'll change the function name to `__containers_from_dump_file` to make things more clear? Happy to talk about this in person. I have a feeling there is something I'm not understanding.

Reply:
Here's the scenario. robnagler is participating in USPAS as a student so is put into the uspas pool. The cname is jupyter-robnagler, but it is running on node xyz1 which is in the uspas pool. robnagler leaves the class and the server is restarted to with him in the community pool, which gets reinitialized as server startup. The pool initialization looks for cname jupyter-robnagler which it finds, but the slot is not in the community pool, that is, host xyz1 is not part of the pool so it can't manage this container. The container on xyz1 should be removed, in fact, because a user can only appear in one pool (at this time). 

previous_slots should probably be called running_containers or something else like that, because slots only exist within the current running process. They are not transferrable, that is, slot_num 35 is not slot_num 35 on restarted server, because you can't assume anything about the slot numbering, since it depends on the configuration.

Reply:
Only test slots of the same pool to make it clear that user's may change pool and the information from the previous pool isn't carried over https://github.com/radiasoft/rsdockerspawner/pull/29/commits/73c1386443375355d61f9e40e3a86e88347e15e2